subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.composite.internal;
+
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.internal.artifacts.ivyservice.projectmodule.BuildTreeLocalComponentProvider;
+import org.gradle.api.internal.artifacts.ivyservice.projectmodule.LocalComponentCache;
+import org.gradle.api.internal.artifacts.ivyservice.projectmodule.LocalComponentProvider;
+import org.gradle.api.internal.project.HoldsProjectState;
+import org.gradle.api.internal.project.ProjectState;
+import org.gradle.api.internal.project.ProjectStateRegistry;
+import org.gradle.internal.Describables;
+import org.gradle.internal.Factory;
+import org.gradle.internal.build.CompositeBuildParticipantBuildState;
+import org.gradle.internal.build.IncludedBuildState;
+import org.gradle.internal.component.local.model.LocalComponentGraphResolveState;
+import org.gradle.internal.component.local.model.LocalComponentGraphResolveStateFactory;
+import org.gradle.internal.component.local.model.LocalComponentMetadata;
+import org.gradle.internal.model.CalculatedValueContainer;
+import org.gradle.internal.model.CalculatedValueContainerFactory;
+
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Default implementation of {@link BuildTreeLocalComponentProvider}.
+ */
+public class DefaultBuildTreeLocalComponentProvider implements BuildTreeLocalComponentProvider, HoldsProjectState {
+
+    private final ProjectStateRegistry projectStateRegistry;
+    private final LocalComponentGraphResolveStateFactory resolveStateFactory;
+    private final LocalComponentCache localComponentCache;
+    private final LocalComponentProvider localComponentProvider;
+
+    private final ConcurrentCache localProjects;
```

Comment:
💭 This seems like a confusing model.  If there are different metadatas based on request, then I would expect a single cache with 2 layers like:

ProjectID -> Request Type (requested within producing build / requested externally) -> Requested Metadata.

It's confusing to me because this type lives at the build tree level, so what a "local" project is doesn't really make sense here.  It only makes sense in the context of a request for metadata, based on the requester.

Reply:
I don't think combining `localProjects` and `foreignProjects` is all that much different  of a concept than what exists currently. If I understand correctly, your idea only adds an extra layer of structure to the data, right?:
`localProjects.get(id)` vs `projects.local.get(id)`

In dependency-management, local is commonly used to refer to any component which is built on the current machine, as opposed to `external` which is used to refer to components sourced from a repository. I think `local` still makes sense at a build-tree level, since it refers to components built on the local machine

Reply:
Updated the names to `originalComponents` and `foreignIdentifiedComponents`

Reply:
This also makes me more comfortable with 2 fields for some reason.  In the original version I thought the same project in both wasn't intuitive.  With the new names focusing on identification I don't find it surprising.